WalletWasabi.Backend/Controllers/BlockchainController.cs

```diff
@@ -140,76 +143,107 @@ private async Task<IEnumerable<string>> GetRawMempoolStringsNoCacheAsync(Cancell
 	[HttpGet("transaction-hexes")]
 	[ProducesResponseType(200)]
 	[ProducesResponseType(400)]
-	public async Task<IActionResult> GetTransactionsAsync([FromQuery, Required] IEnumerable<string> transactionIds, CancellationToken cancellationToken)
+	public async Task<IActionResult> GetTransactionsAsync([FromQuery, Required] List<string> transactionIds, CancellationToken cancellationToken)
 	{
-		var maxTxToRequest = 10;
-		if (transactionIds.Count() > maxTxToRequest)
+		const int MaxTxToRequest = 10;
+		int requestCount = transactionIds.Count;
+
+		if (requestCount > MaxTxToRequest)
 		{
-			return BadRequest($"Maximum {maxTxToRequest} transactions can be requested.");
+			return BadRequest($"Maximum {MaxTxToRequest} transactions can be requested.");
 		}
 
-		var parsedIds = new List<uint256>();
+		uint256[] parsedTxIds;
+
+		// Make sure TXIDs are not malformed.
 		try
 		{
-			// Remove duplicates, do not use Distinct(), order is not guaranteed.
-			foreach (var txid in transactionIds.Select(x => new uint256(x)))
-			{
-				if (!parsedIds.Contains(txid))
-				{
-					parsedIds.Add(txid);
-				}
-			}
+			parsedTxIds = transactionIds.Select(x => new uint256(x)).ToArray();
 		}
 		catch
 		{
 			return BadRequest("Invalid transaction Ids.");
 		}
 
-		try
+		if (requestCount == 1)
+		{
+			var singleTx = parsedTxIds.Single();
+			var tx = await GetTransactionAsync(singleTx, cancellationToken);
+			return Ok(new[] { tx });
+		}
+
+		HashSet<uint256> txIdsRetrieve = [];
+		Task<Transaction>?[] tasks = new Task<Transaction>?[requestCount];
+
+		// Fill transactions we already have in cache. Array can look like [null, tx1, null, tx2, etc.].
+		for (int i = 0; i < requestCount; i++)
 		{
-			var hexes = new Dictionary<uint256, string>();
-			List<uint256> missingTxs = new();
-			lock (TransactionHexCacheLock)
+			uint256 txId = parsedTxIds[i];
+			string cacheKey = GetCacheKeyForTransaction(txId);
+
+			if (Cache.TryGetValue(cacheKey, out TaskCompletionSource<Transaction>? tcs))
 			{
-				foreach (var txid in parsedIds)
-				{
-					if (TransactionHexCache.TryGetValue(txid, out string? hex))
-					{
-						hexes.Add(txid, hex);
-					}
-					else
-					{
-						missingTxs.Add(txid);
-					}
-				}
+				tasks[i] = tcs!.Task;
 			}
-
-			if (missingTxs.Count != 0)
+			else
 			{
-				foreach (var tx in await RpcClient.GetRawTransactionsAsync(missingTxs, cancellationToken))
-				{
-					string hex = tx.ToHex();
-					hexes.Add(tx.GetHash(), hex);
-
-					lock (TransactionHexCacheLock)
-					{
-						if (TransactionHexCache.TryAdd(tx.GetHash(), hex) && TransactionHexCache.Count >= 1000)
-						{
-							TransactionHexCache.Remove(TransactionHexCache.Keys.First());
-						}
-					}
-				}
+				txIdsRetrieve.Add(txId);
 			}
+		}
+
+		// Ask to get missing transactions over RPC.
+		Task<Dictionary<uint256, Transaction>> rpcBatchTask = Task.Run(async () =>
```

Comment:
If we are going for edge performance we should assume tons of requests. What happens if since the last cycle another request arrives requiring the same transactions?